cmd/snap-bootstrap/initramfs_systemd_mount.go

```diff
@@ -0,0 +1,187 @@
+	whereEscaped := systemd.EscapeUnitNamePath(where)
+	unitName := whereEscaped + ".mount"
+
+	args := []string{what, where, "--no-pager", "--no-ask-password"}
+	if opts.IsTmpfs {
+		args = append(args, "--type=tmpfs")
+	}
+
+	if opts.NeedsFsck {
+		// note that with the --fsck=yes argument, systemd will block starting
+		// the mount unit on a new systemd-fsck@<what> unit that will run the
+		// fsck, so we don't need to worry about waiting for that to finish in
+		// the case where we are supposed to wait (which is the default for this
+		// function)
+		args = append(args, "--fsck=yes")
+	}
+
+	// Under all circumstances that we use systemd-mount here from
+	// snap-bootstrap, it is expected to be okay to block waiting for the unit
+	// to be started and become active, because snap-bootstrap is, by design,
+	// expected to run as late as possible in the initramfs, and so any
+	// dependencies there might be in systemd creating and starting these mount
+	// units should already be ready and so we will not block forever. If
+	// however there was something going on in systemd at the same time that the
+	// mount unit depended on, we could hit a deadlock blocking as systemd will
+	// not enqueue this job until it's dependencies are ready, and so if those
+	// things depend on this mount unit we are stuck. The solution to this
+	// situation is to make snap-bootstrap run as late as possible before
+	// mounting things.
+	// However, we leave in the option to not block if there is ever a reason
+	// we need to do so.
+	if opts.NoWait {
+		args = append(args, "--no-block")
+	}
+
+	// note that we do not currently parse any output from systemd-mount, but if
+	// we ever do, take special care surrounding the debug output that systemd
+	// outputs with the "debug" kernel command line present (or equivalently the
+	// SYSTEMD_LOG_LEVEL=debug env var) which will add lots of additional output
+	// to stderr from systemd commands
+	out, err := exec.Command("systemd-mount", args...).CombinedOutput()
+	if err != nil {
+		return osutil.OutputErr(out, err)
+	}
+
+	// if it should survive pivot_root() then we need to add overrides for this
+	// unit to /run/systemd units
+	if !opts.EphemeralInitramfsMount {
+		// to survive the pivot_root, we need to make the mount units depend on
+		// all of the various initrd special targets by adding runtime conf
+		// files there
+		// note we could do this statically in the initramfs main filesystem
+		// layout, but that means that changes to snap-bootstrap would block on
+		// waiting for those files to be added before things works here, this is
+		// a more flexible strategy that puts snap-bootstrap in control
+		for _, initrdUnit := range []string{
+			"initrd.target",
+			"initrd-fs.target",
+			"initrd-switch-root.target",
+			"local-fs.target",
+		} {
+			targetDir := filepath.Join(dirs.GlobalRootDir, "/run/systemd/system", initrdUnit+".d")
+			err := os.MkdirAll(targetDir, 0755)
+			if err != nil {
+				return err
```

Comment:
the doc comment of the function should maybe clarify that is expected that this function can error leaving the mount in place